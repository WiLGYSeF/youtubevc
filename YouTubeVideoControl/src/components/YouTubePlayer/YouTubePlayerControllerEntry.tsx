import React, { useState } from 'react';
import '../../css/style.min.css';

function YouTubePlayerControllerEntry() {
  return (
    <div>
    </div>
  );
}

export default YouTubePlayerControllerEntry;
