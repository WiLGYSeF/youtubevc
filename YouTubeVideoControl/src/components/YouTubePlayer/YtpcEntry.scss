.entry {
  border: 1px solid #aaa;
  display: flex;
  height: 25px;
  justify-content: space-between;
  padding: 2px 4px;

  .buttons {
    display: flex;
  }

  .delete, .edit {
    background-color: #fff;
    border: 1px solid #aaa;
    border-radius: 4px;
    color: #000;
    height: 20px;
    user-select: none;
    width: 20px;

    &:hover {
      background-color: #ddd;
    }
  }
}
