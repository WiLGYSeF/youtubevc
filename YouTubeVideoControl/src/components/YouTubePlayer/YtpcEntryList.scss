.entry-list {
  background-color: #444;
  border: 1px solid #aaa;
  border-radius: 6px;
  color: #ddd;
  height: 200px;
  overflow: auto;
}
