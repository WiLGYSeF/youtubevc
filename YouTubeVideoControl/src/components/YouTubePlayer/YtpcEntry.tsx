import React, { useEffect, useReducer } from 'react';

import YouTubePlayerControllerEntry from 'objects/YtpcEntry/YouTubePlayerControllerEntry';

import './YtpcEntry.scss';

interface YtpcEntryProps {
  entry: YouTubePlayerControllerEntry;
  deleteEntry(entry: YouTubePlayerControllerEntry): void;
  editEntry(entry: YouTubePlayerControllerEntry): void;
}

const UPDATE_INTERVAL = 100;

function YtpcEntry(props: YtpcEntryProps) {
  const [, forceUpdate] = useReducer((x) => x + 1, 0);

  useEffect(() => {
    // TODO: replace with something better
    const timeout = setTimeout(forceUpdate, UPDATE_INTERVAL);
    return () => {
      clearTimeout(timeout);
    };
  });

  return (
    <div
      className="entry"
      data-at-time={props.entry.atTime}
      data-control-type={props.entry.controlType}
      data-entry={JSON.stringify(props.entry)}
    >
      <span>{`${props.entry}`}</span>
      <div className="buttons">
        <div
          className="edit"
          onClick={() => {
            props.editEntry(props.entry);
          }}
        >
          e
        </div>
        <div
          className="delete"
          onClick={() => {
            props.deleteEntry(props.entry);
          }}
        >
          x
        </div>
      </div>
    </div>
  );
}

export default YtpcEntry;
