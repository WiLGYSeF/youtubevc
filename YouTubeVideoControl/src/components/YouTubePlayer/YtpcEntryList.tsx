import React from 'react';

import YouTubePlayerControllerEntry from 'objects/YtpcEntry/YouTubePlayerControllerEntry';
import YtpcEntry from './YtpcEntry';
import YtpcEntryBar from './YtpcEntryBar';

import styles from './YtpcEntryList.module.scss';

interface YtpcEntryListProps {
  entries: YouTubePlayerControllerEntry[];
  barIndex?: number;
  deleteEntry(entry: YouTubePlayerControllerEntry): void;
  editEntry(entry: YouTubePlayerControllerEntry): void;
}

function YtpcEntryList(props: YtpcEntryListProps) {
  const elements = props.entries.map((e, i) => (
    <YtpcEntry
      key={e.getKey()}
      entry={e}
      deleteEntry={props.deleteEntry}
      editEntry={props.editEntry}
    />
  ));

  if (props.barIndex !== undefined) {
    elements.splice(
      props.barIndex,
      0, (
        <YtpcEntryBar key="entrybar" />
      ),
    );
  }

  return (
    <div className={styles['entry-list']}>
      {elements}
    </div>
  );
}

export function getEntries(container: Element): Element[] {
  return Array.from(container.querySelectorAll('.entry'));
}

export default YtpcEntryList;
