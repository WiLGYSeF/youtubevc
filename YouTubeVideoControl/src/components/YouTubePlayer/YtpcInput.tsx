import React, { useEffect, useState } from 'react';
import { YouTubePlayer } from 'youtube-player/dist/types';

import { ControlType } from '../../objects/YtpcEntry/YouTubePlayerControllerEntry';
import YtpcControlSelect from './YtpcControlSelect';
import YtpcAdd from './YtpcAdd';
import YtpcInputGoto from './YtpcInputGoto';
import TimestampInput from '../common/TimestampInput';
import secondsToTimestamp from '../../utils/secondsToTimestamp';

import '../../css/style.min.css';

interface YtpcInputProps {
  ytPlayer?: YouTubePlayer;
  is360Video: boolean;

  controlInputType: ControlType;

  createEntry(type: ControlType, atTime: number, state: object): void;
}

function YtpcInput(props: YtpcInputProps) {
  const [atTime, setAtTime] = useState(0);
  const [nowTime, setNowTime] = useState(0);
  const [controlInput, setControlInput] = useState(() => YtpcInputGoto);
  const [controlInputType, setControlInputType] = useState(props.controlInputType);
  const [controlInputState, setControlInputState] = useState<object>({});

  useEffect(() => {
    setControlInputType(props.controlInputType);
  }, [props.controlInputType]);

  return (
    <div className="input">
      <div className="entry-creation">
        <span>At </span>
        <span
          className="now-time"
          title="Click to use current time in video"
          onClick={() => {
            if (props.ytPlayer) {
              const curTime = Math.floor(props.ytPlayer.getCurrentTime() * 100) / 100;
              setNowTime(curTime);
            }
          }}
        >
          *
        </span>
        <TimestampInput
          defaultValue={secondsToTimestamp(nowTime)}
          setTime={setAtTime}
        />
        <span>, </span>
        <YtpcControlSelect
          is360Video={props.is360Video}
          controlInputType={controlInputType}
          setControlInput={(type: ControlType, component?: any) => {
            setControlInput(component);
            setControlInputType(type);
          }}
        />
        {React.createElement(controlInput, {
          setControlInputState,
          playbackRates: props.ytPlayer?.getAvailablePlaybackRates(),
        })}
      </div>
      <YtpcAdd createEntry={() => props.createEntry(controlInputType, atTime, controlInputState)} />
    </div>
  );
}

export default YtpcInput;
