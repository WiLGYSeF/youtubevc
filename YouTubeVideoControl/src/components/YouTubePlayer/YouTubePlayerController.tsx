import React, { useEffect, useState } from 'react';
import { YouTubePlayer } from 'youtube-player/dist/types';
import PlayerStates from 'youtube-player/dist/constants/PlayerStates';

import Ytpc360Entry, { YouTubePlayer360 } from '../../objects/YtpcEntry/Ytpc360Entry';
import YouTubePlayerControllerEntry, { ControlType } from '../../objects/YtpcEntry/YouTubePlayerControllerEntry';
import YtpcClear from './YtpcClear';
import YtpcCopyLink from './YtpcCopyLink';
import YtpcEntryList from './YtpcEntryList';
import YtpcExport, { ExportType } from './YtpcExport';
import YtpcImport from './YtpcImport';
import YtpcInput from './YtpcInput';
import EntryBuilder from '../../objects/YtpcEntry/EntryBuilder';
import Coroutine from '../../utils/coroutine';

import '../../css/style.min.css';

import YtpcGotoEntry from '../../objects/YtpcEntry/YtpcGotoEntry';
import YtpcLoopEntry from '../../objects/YtpcEntry/YtpcLoopEntry';
import YtpcPauseEntry from '../../objects/YtpcEntry/YtpcPauseEntry';
import YtpcPlaybackRateEntry from '../../objects/YtpcEntry/YtpcPlaybackRateEntry';
import YtpcVolumeEntry from '../../objects/YtpcEntry/YtpcVolumeEntry';

interface YouTubePlayerControllerProps {
  ytPlayer?: YouTubePlayer;
  entries?: string | null;
}

const EVENT_ONSTATECHANGE = 'onStateChange';

const TIME_DIFF_MAX = 0.1;

function addEntry(entries: YouTubePlayerControllerEntry[], entry: YouTubePlayerControllerEntry): void {
  const existingEntryIndex = entries.findIndex(
    (e) => e.atTime === entry.atTime && e.controlType === entry.controlType,
  );

  if (existingEntryIndex === -1) {
    entries.push(entry);
  } else {
    entries[existingEntryIndex] = entry;
  }

  entries.sort(
    (a: YouTubePlayerControllerEntry, b: YouTubePlayerControllerEntry): number => a.atTime - b.atTime,
  );
}

function getVideoIdByUrl(url: string): string | null {
  try {
    const urlObj = new URL(url);
    return urlObj.searchParams.get('v');
  } catch {
    return null;
  }
}

function playerHas360Video(player: YouTubePlayer360): boolean {
  return Object.keys(player.getSphericalProperties()).length > 0;
}

function YouTubePlayerController(props: YouTubePlayerControllerProps) {
  const [entries, setEntries] = useState<YouTubePlayerControllerEntry[]>([
    new YtpcGotoEntry(0, 1),
    new YtpcGotoEntry(3, 6),
    new YtpcPauseEntry(30, 1),
    new YtpcPlaybackRateEntry(31, 2),
    new YtpcVolumeEntry(32, 30, 5),
    new YtpcVolumeEntry(33, 80),
    new Ytpc360Entry(33, {
      yaw: 1, pitch: 2, roll: 3, fov: 5,
    }),
    new Ytpc360Entry(34, {
      yaw: 5, pitch: 2, roll: 3, fov: 5,
    }, 2),
    new YtpcLoopEntry(7 * 60 + 51, 3 * 60 + 25),
    new YtpcLoopEntry(7 * 60 + 52, 3 * 60 + 25, 3),
  ]);
  const [barIndex, setBarIndex] = useState(0);
  const [is360Video, setIs360Video] = useState(false);
  const [atTime, setAtTime] = useState(0);
  const [controlInputType, setControlInputType] = useState(ControlType.Goto);
  const [controlInputState, setControlInputState] = useState<object>({});

  useEffect(() => {
    const onStateChange = (e: CustomEvent) => {
      const state: PlayerStates = (e as any).data;
      const has360Video = playerHas360Video(props.ytPlayer as YouTubePlayer360);

      if (is360Video !== has360Video) {
        setIs360Video(has360Video);
      }
    };

    props.ytPlayer?.addEventListener(EVENT_ONSTATECHANGE, onStateChange);

    return () => {
      props.ytPlayer?.removeEventListener(EVENT_ONSTATECHANGE, onStateChange);
    };
  });

  useEffect(() => {
    if (!props.entries) {
      return;
    }

    try {
      const parsedEntries: YouTubePlayerControllerEntry[] = [];

      JSON.parse(props.entries).forEach(
        (o) => addEntry(parsedEntries, EntryBuilder.buildEntry(o.controlType, o.atTime, o))
      );
      setEntries(parsedEntries);
    } catch (exc) {
      console.error(exc);
    }
  }, [props.entries]);

  useEffect(() => {
    let lastTime = props.ytPlayer?.getCurrentTime() ?? 0;

    const routine = new Coroutine(() => {
      if (!props.ytPlayer) {
        return;
      }

      const curTime = props.ytPlayer.getCurrentTime();

      let idx = 0;
      let lastMatchingIdx = -1;

      for (const entry of entries) {
        if (entry.atTime < curTime) {
          if (entry.atTime >= Math.max(lastTime, curTime - TIME_DIFF_MAX)) {
            entry.performAction(props.ytPlayer, curTime);
          }
          lastMatchingIdx = idx;
        }
        idx += 1;
      }

      setBarIndex(lastMatchingIdx + 1);

      lastTime = curTime;
    });
    routine.start();

    return () => {
      routine.stop();
    };
  }, [props.ytPlayer, entries]);

  const deleteEntry = (entry: YouTubePlayerControllerEntry): void => {
    setEntries([...entries.filter((e) => e !== entry)]);
  };

  const editEntry = (entry: YouTubePlayerControllerEntry): void => {
    setAtTime(entry.atTime);
    setControlInputType(entry.controlType);
    setControlInputState(entry.getState());
  };

  const clearEntries = (): void => {
    setEntries([]);
  };

  return (
    <div className="yt-controller">
      <div className="left">
        <YtpcInput
          ytPlayer={props.ytPlayer}
          is360Video={is360Video}
          atTime={atTime}
          controlInputType={controlInputType}
          controlInputState={controlInputState}
          createEntry={(type: ControlType, atTime: number, state: object) => {
            addEntry(entries, EntryBuilder.buildEntry(type, atTime, state));
            setEntries(entries);
          }}
          setControlInputState={setControlInputState}
        />

        <YtpcEntryList
          entries={entries}
          barIndex={barIndex}
          deleteEntry={deleteEntry}
          editEntry={editEntry}
        />

        <div>
          <YtpcClear clearEntries={clearEntries} />
          <YtpcImport addEntry={addEntry} setEntries={setEntries} />
          <YtpcExport
            filename={`youtubevc-${getVideoIdByUrl(props.ytPlayer?.getVideoUrl() ?? '')}.txt`}
            entries={entries}
            exportType={ExportType.Text}
          />
          <YtpcCopyLink videoId={getVideoIdByUrl(props.ytPlayer?.getVideoUrl() ?? '')} entries={entries} />
        </div>
      </div>
      <div className="right" />
    </div>
  );
}

export default YouTubePlayerController;
