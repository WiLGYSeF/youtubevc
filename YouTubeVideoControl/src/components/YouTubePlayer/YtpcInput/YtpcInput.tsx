import React, { useEffect, useState } from 'react';
import { YouTubePlayer } from 'youtube-player/dist/types';

import TimestampInput from 'components/common/TimestampInput/TimestampInput';
import { YtpcEntryState } from 'objects/YtpcEntry/YouTubePlayerControllerEntry';
import round from 'utils/round';
import YtpcAdd from '../YtpcAdd';
import { YtpcControlInput } from './YtpcControlInput';
import YtpcControlSelect, { controlTypeToComponent } from './YtpcControlSelect';

import './YtpcInput.scss';

interface YtpcInputProps {
  ytPlayer?: YouTubePlayer;
  is360Video: boolean;

  defaultState: YtpcEntryState;
  setDefaultState(state: YtpcEntryState): void;
  entryState: YtpcEntryState;
  setEntryState(state: YtpcEntryState): void;

  createEntry(state: YtpcEntryState): void;
}

function YtpcInput(props: YtpcInputProps) {
  const component = controlTypeToComponent(props.defaultState.controlType);
  const [controlInput, setControlInput] = useState(() => component);

  const createComponent = (input: YtpcControlInput) => React.createElement(controlInput, input);

  useEffect(() => {
    setControlInput(() => component);
  }, [props.defaultState.controlType]);

  return (
    <div className="input">
      <div className="entry-creation">
        <span>At </span>
        <span
          className="now-time"
          title="Click to use current time in video"
          onClick={() => {
            if (props.ytPlayer) {
              const state = { ...props.defaultState };

              state.atTime = round(props.ytPlayer.getCurrentTime(), 2);
              props.setDefaultState(state);
            }
          }}
        >
          *
        </span>
        <span className="at-time">
          <TimestampInput
            defaultValue={props.defaultState.atTime}
            onChange={(seconds: number) => {
              const state = { ...props.entryState };
              state.atTime = seconds;
              props.setEntryState(state);
            }}
          />
        </span>
        <span>, </span>
        <span className="control-select">
          <YtpcControlSelect
            is360Video={props.is360Video}
            defaultControlType={props.defaultState.controlType}
            setControlInput={(type, component) => {
              const state = { ...props.entryState };
              state.controlType = type;

              setControlInput(() => component);
              props.setEntryState(state);
            }}
          />
        </span>
        <span className="control-input">
          {createComponent({
            entryState: props.entryState,
            defaultState: props.defaultState,
            setEntryState: props.setEntryState,
            playbackRates: props.ytPlayer?.getAvailablePlaybackRates(),
          })}
        </span>
      </div>
      <span className="add">
        <YtpcAdd createEntry={() => props.createEntry(props.entryState)} />
      </span>
    </div>
  );
}

export default YtpcInput;
