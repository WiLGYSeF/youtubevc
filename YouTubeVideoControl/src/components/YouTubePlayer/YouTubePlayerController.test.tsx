import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import YouTubePlayerControllerEntry, { ControlType } from 'objects/YtpcEntry/YouTubePlayerControllerEntry';
import YtpcGotoEntry from 'objects/YtpcEntry/YtpcGotoEntry';
import YtpcLoopEntry from 'objects/YtpcEntry/YtpcLoopEntry';
import { secondsToTimestamp } from 'utils/timestr';
import { PLAYBACK_RATES } from 'utils/youtube';
import { YouTubePlayer } from 'youtube-player/dist/types';
import YouTubePlayerController, {
  performEntryActions,
  addEntry, filterLoopEntries, getInputs, getRandomLoopEntry,
} from './YouTubePlayerController';
import { getInputsByControl } from './YtpcInput/YtpcInput';
import { YtpcInputLoopInputs } from './YtpcInput/YtpcInputLoop';
import { YtpcInputGotoInputs } from './YtpcInput/YtpcInputGoto';
import { getInputs as entryGetInputs } from './YtpcEntry';
import { getEntries } from './YtpcEntryList';

function expectAscendingOrder(arr: any[], cmp: (a: any, b: any) => number): void {
  for (let i = 1; i < arr.length; i += 1) {
    expect(cmp(arr[i - 1], arr[i]) <= 0).toBeTruthy();
  }
}

function getEntryList(container: HTMLElement): HTMLElement {
  return container.querySelector('[data-testid="entry-list"]')!;
}

describe('YouTubePlayerController', () => {
  describe('addEntry', () => {
    it('adds entries in order', () => {
      const entries = [
        new YtpcGotoEntry(1, 12),
        new YtpcGotoEntry(2, 34),
        new YtpcLoopEntry(5, 3),
        new YtpcGotoEntry(10, 78),
      ];

      addEntry(entries, new YtpcGotoEntry(3, 0));
      addEntry(entries, new YtpcLoopEntry(11, 0));

      expect(entries.length).toEqual(6);
      expectAscendingOrder(
        entries,
        (a: YouTubePlayerControllerEntry, b: YouTubePlayerControllerEntry) => a.atTime - b.atTime,
      );
    });

    it('modifies existing entries', () => {
      const atTime = 5;
      const entries = [
        new YtpcGotoEntry(1, 12),
        new YtpcGotoEntry(2, 34),
        new YtpcGotoEntry(atTime, 3),
        new YtpcLoopEntry(10, 5),
      ];

      expect(
        (entries.find(
          (e) => e.controlType === ControlType.Goto && e.atTime === atTime,
        ) as YtpcGotoEntry).gotoTime,
      ).toEqual(3);

      addEntry(entries, new YtpcGotoEntry(atTime, 1));
      addEntry(entries, new YtpcLoopEntry(atTime, 0));

      expect(entries.length).toEqual(5);
      expect(
        (entries.find(
          (e) => e.controlType === ControlType.Goto && e.atTime === atTime,
        ) as YtpcGotoEntry
        ).gotoTime,
      ).toEqual(1);
      expectAscendingOrder(
        entries,
        (a: YouTubePlayerControllerEntry, b: YouTubePlayerControllerEntry) => a.atTime - b.atTime,
      );
    });
  });

  describe('getRandomLoopEntry', () => {
    const entries = [
      new YtpcLoopEntry(20, 10, 10),
      new YtpcLoopEntry(37, 21, 3),
      new YtpcGotoEntry(37.5, 0),
      new YtpcLoopEntry(42, 38, 2),
      new YtpcLoopEntry(51, 44, 3),
      new YtpcGotoEntry(53, 0),
      new YtpcLoopEntry(60, 55, 1),
    ];

    it('filters only loop entries', () => {
      for (const type of filterLoopEntries(entries).map((e) => e.controlType)) {
        expect(type).toEqual(ControlType.Loop);
      }
    });

    it.each([
      [0.5, false, 42],
      [0.15, false, 20],
      [0.85, false, 60],
      [0.5, true, 20],
      [0.15, true, 20],
      [0.85, true, 51],
    ])(
      'gets random entries',
      (random: number, weighted: boolean, expected: number) => {
        const randomMock = jest.spyOn(Math, 'random').mockImplementation(() => random);

        const entry = getRandomLoopEntry(entries, weighted);

        expect(entry.controlType).toEqual(ControlType.Loop);
        expect(entry.atTime).toEqual(expected);

        randomMock.mockRestore();
      },
    );
  });

  describe('performEntryActions', () => {
    const ytPlayer = {
      seekTo: jest.fn(),
    } as unknown as YouTubePlayer;

    it('performs entry actions', () => {
      const performActionMock = jest.spyOn(YtpcGotoEntry.prototype, 'performAction')
        .mockImplementation();

      const entries = [
        new YtpcGotoEntry(4, 10),
        new YtpcGotoEntry(15, 0),
      ];

      let lastMatchingIdx = 0;

      const doTest = (curTime: number, lastTime: number) => {
        performActionMock.mockClear();
        lastMatchingIdx = performEntryActions(ytPlayer, entries, curTime, lastTime, false, false);
      };

      for (let i = 0; i < entries[0].atTime; i += 0.1) {
        doTest(i - 0.1, i);
        expect(lastMatchingIdx).toEqual(-1);
        expect(performActionMock).toHaveBeenCalledTimes(0);
      }

      doTest(entries[0].atTime, entries[0].atTime - 0.1);
      expect(lastMatchingIdx).toEqual(0);
      expect(performActionMock).toHaveBeenCalledTimes(1);

      for (let i = 0; i < entries.length; i += 1) {
        doTest(entries[i].atTime + 0.005, entries[i].atTime - 0.005);
        expect(lastMatchingIdx).toEqual(i);
        expect(performActionMock).toHaveBeenCalledTimes(1);
      }

      doTest(entries[1].atTime - 0.005, entries[1].atTime - 0.01);
      expect(lastMatchingIdx).toEqual(0);
      expect(performActionMock).toHaveBeenCalledTimes(0);

      performActionMock.mockRestore();
    });

    it('performs multiple actions', () => {
      const performActionMock = jest.spyOn(YtpcGotoEntry.prototype, 'performAction')
        .mockImplementation();

      const entries = [
        new YtpcGotoEntry(15, 1),
        new YtpcGotoEntry(15.1, 0),
      ];

      const lastMatchingIdx = performEntryActions(
        ytPlayer,
        entries,
        entries[1].atTime,
        entries[0].atTime,
        false,
        false
      );

      expect(lastMatchingIdx).toEqual(1);
      expect(performActionMock).toHaveBeenCalledTimes(2);
      expect(performActionMock.mock.instances).toEqual(entries);

      performActionMock.mockRestore();
    });

    it('performs entry actions with loop shuffle', () => {
      const entries = [
        new YtpcLoopEntry(15, 0),
      ];

      let lastMatchingIdx = 0;

      const doTest = (curTime: number, lastTime: number) => {
        (ytPlayer.seekTo as jest.Mock).mockClear();
        lastMatchingIdx = performEntryActions(ytPlayer, entries, curTime, lastTime, true, false);
      };

      for (let i = 0; i < entries[0].atTime; i += 0.1) {
        doTest(i - 0.1, i);
        expect(lastMatchingIdx).toEqual(-1);
        expect(ytPlayer.seekTo).toHaveBeenCalledTimes(0);
      }

      doTest(entries[0].atTime, entries[0].atTime - 0.1);
      expect(lastMatchingIdx).toEqual(0);
      expect(ytPlayer.seekTo).toHaveBeenCalledTimes(1);

      for (let i = 0; i < entries.length; i += 1) {
        doTest(entries[i].atTime + 0.005, entries[i].atTime - 0.005);
        expect(lastMatchingIdx).toEqual(i);
        expect(ytPlayer.seekTo).toHaveBeenCalledTimes(1);
      }

      (ytPlayer.seekTo as jest.Mock).mockRestore();
    });
  });

  describe('YouTubePlayerController', () => {
    const getEntryObject = (
      entry: HTMLElement,
    ): YouTubePlayerControllerEntry => JSON.parse(entry.dataset.entry!);

    const ytPlayer = {
      addEventListener: jest.fn(),
      getAvailablePlaybackRates: jest.fn(() => PLAYBACK_RATES),
      getCurrentTime: jest.fn(),
      getVideoUrl: jest.fn(),
      removeEventListener: jest.fn(),
      seekTo: jest.fn(),
    } as unknown as YouTubePlayer;

    it('adds entries', () => {
      const { container } = render(<YouTubePlayerController
        ytPlayer={ytPlayer}
        entries=""
        loopShuffle={false}
        shuffleWeight={false}
      />);

      const { input } = getInputs(container);

      userEvent.clear(input.atTime.input);
      userEvent.type(input.atTime.input, '12:34');

      userEvent.selectOptions(input.controlSelect.select, ControlType.Loop);

      const {
        loopBackTo,
        forever,
        loopCount,
      } = getInputsByControl(input.controlInput, ControlType.Loop) as YtpcInputLoopInputs;

      userEvent.clear(loopBackTo.input);
      userEvent.type(loopBackTo.input, '56');

      userEvent.click(forever.checkbox);

      userEvent.clear(loopCount.input);
      userEvent.type(loopCount.input, '4');

      userEvent.click(input.add.add);

      const entryList = getEntryList(container);
      const entries = getEntries(entryList) as HTMLElement[];

      expect(entries.length).toEqual(1);
      expect(JSON.parse(entries[0].dataset.entry!)).toEqual({
        atTime: 754,
        controlType: ControlType.Loop,
        loopBackTo: 56,
        loopCount: 4,
        loopNum: 0,
      });
    });

    it('deletes entries', () => {
      const expected = [
        {
          atTime: 0,
          controlType: ControlType.Goto,
          gotoTime: 15,
        },
        {
          atTime: 10,
          controlType: ControlType.Goto,
          gotoTime: 25,
        },
        {
          atTime: 123,
          controlType: ControlType.Goto,
          gotoTime: 4,
        },
      ];

      const { container } = render(<YouTubePlayerController
        ytPlayer={ytPlayer}
        entries={JSON.stringify(expected)}
        loopShuffle={false}
        shuffleWeight={false}
      />);

      const entryList = container.querySelector('.entry-list')!;
      let entries = getEntries(entryList) as HTMLElement[];

      expect(entries.length).toEqual(3);
      expect(entries.map(getEntryObject)).toEqual(expected);

      const { eDelete } = entryGetInputs(entries[1]);

      userEvent.click(eDelete);
      expected.splice(1, 1);

      entries = getEntries(entryList) as HTMLElement[];

      expect(entries.length).toEqual(2);
      expect(entries.map(getEntryObject)).toEqual(expected);
    });

    it('edits an entry', () => {
      const expected = [
        {
          atTime: 0,
          controlType: ControlType.Goto,
          gotoTime: 15,
        },
        {
          atTime: 10,
          controlType: ControlType.Goto,
          gotoTime: 25,
        },
        {
          atTime: 15,
          controlType: ControlType.Loop,
          loopBackTo: 0,
          loopCount: 3,
        },
        {
          atTime: 123,
          controlType: ControlType.Goto,
          gotoTime: 4,
        },
        {
          atTime: 124,
          controlType: ControlType.Loop,
          loopBackTo: 20,
          loopCount: -1,
        },
      ];

      const tests = [
        {
          index: 1,
          controlType: ControlType.Goto,
        },
        {
          index: 3,
          controlType: ControlType.Goto,
        },
        {
          index: 2,
          controlType: ControlType.Loop,
        },
        {
          index: 4,
          controlType: ControlType.Loop,
        },
      ];

      const { container } = render(<YouTubePlayerController
        ytPlayer={ytPlayer}
        entries={JSON.stringify(expected)}
        loopShuffle={false}
        shuffleWeight={false}
      />);

      const { input } = getInputs(container);

      const entryList = container.querySelector('.entry-list')!;
      const entries = getEntries(entryList) as HTMLElement[];

      for (const test of tests) {
        const eEntry = entries[test.index];
        const entry = getEntryObject(eEntry);
        const { edit } = entryGetInputs(eEntry);

        userEvent.click(edit);

        expect(input.atTime.input.value).toEqual(secondsToTimestamp(entry.atTime));
        expect(input.controlSelect.select.value).toEqual(test.controlType);

        switch (test.controlType) {
          case ControlType.Goto: {
            const gotoEntry = entry as YtpcGotoEntry;
            const gotoInput = getInputsByControl(
              input.controlInput,
              ControlType.Goto
            ) as YtpcInputGotoInputs;

            expect(gotoInput.gotoTime.input.value)
              .toEqual(secondsToTimestamp(gotoEntry.gotoTime));
            break;
          }
          case ControlType.Loop: {
            const loopEntry = entry as YtpcLoopEntry;
            const loopInput = getInputsByControl(
              input.controlInput,
              ControlType.Loop
            ) as YtpcInputLoopInputs;

            expect(loopInput.loopBackTo.input.value)
              .toEqual(secondsToTimestamp(loopEntry.loopBackTo));
            expect(loopInput.forever.checkbox.checked).toEqual(loopEntry.loopCount < 0);
            expect(loopInput.loopCount.input.value).toEqual(loopEntry.loopCount.toString());
            break;
          }
          default:
            throw new Error('not implemented');
        }
      }
    });

    it('clears entries', () => {
      const expected = [
        {
          atTime: 0,
          controlType: ControlType.Goto,
          gotoTime: 15,
        },
        {
          atTime: 10,
          controlType: ControlType.Goto,
          gotoTime: 25,
        },
        {
          atTime: 123,
          controlType: ControlType.Goto,
          gotoTime: 4,
        },
      ];

      const { container } = render(<YouTubePlayerController
        ytPlayer={ytPlayer}
        entries={JSON.stringify(expected)}
        loopShuffle={false}
        shuffleWeight={false}
      />);

      const entryList = container.querySelector('.entry-list')!;
      let entries = getEntries(entryList) as HTMLElement[];

      const { clear } = getInputs(container);

      userEvent.click(clear.clear);

      entries = getEntries(entryList) as HTMLElement[];

      expect(entries.length).toEqual(0);
    });
  });
});
