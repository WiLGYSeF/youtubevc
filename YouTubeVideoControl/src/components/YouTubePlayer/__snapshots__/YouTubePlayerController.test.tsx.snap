// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YtpcEntryList renders entries 1`] = `
<div
  className="entry-list"
>
  <div
    className="entry"
    data-at-time={0}
    data-control-type="goto"
    data-entry="{\\"controlType\\":\\"goto\\",\\"atTime\\":0,\\"gotoTime\\":0}"
  >
    <span>
      At 00:00, go to 00:00
    </span>
    <div
      className="buttons"
    >
      <div
        className="edit"
        onClick={[Function]}
      >
        e
      </div>
      <div
        className="delete"
        onClick={[Function]}
      >
        x
      </div>
    </div>
  </div>
  <div
    className="entry"
    data-at-time={5}
    data-control-type="goto"
    data-entry="{\\"controlType\\":\\"goto\\",\\"atTime\\":5,\\"gotoTime\\":0}"
  >
    <span>
      At 00:05, go to 00:00
    </span>
    <div
      className="buttons"
    >
      <div
        className="edit"
        onClick={[Function]}
      >
        e
      </div>
      <div
        className="delete"
        onClick={[Function]}
      >
        x
      </div>
    </div>
  </div>
  <div
    className="entry"
    data-at-time={8}
    data-control-type="goto"
    data-entry="{\\"controlType\\":\\"goto\\",\\"atTime\\":8,\\"gotoTime\\":0}"
  >
    <span>
      At 00:08, go to 00:00
    </span>
    <div
      className="buttons"
    >
      <div
        className="edit"
        onClick={[Function]}
      >
        e
      </div>
      <div
        className="delete"
        onClick={[Function]}
      >
        x
      </div>
    </div>
  </div>
</div>
`;
